### PR TITLE
[Core] Clean up base LinearSolver

### PR DESCRIPTION
**Pull request by loumalouomega**
Cleaning up

### CODE REVIEW

**Comment by pooyan-dadvand** [comment]
The warning shouldn't be given by the who call the `IsConsistent` method instead of inside it?

**Comment by philbucher** [comment]
> The warning shouldn't be given by the who call the `IsConsistent` method instead of inside it?

makes sense to me

**Comment by loumalouomega** [comment]
> 
> 
> The warning shouldn't be given by the who call the `IsConsistent` method instead of inside it?

No, because the method only returns a boolean, you don't know that fails. What I understand is that the warning is launch by the method, and the linear solver decide what to do with that information. Maybe throw an error, maybe resize the system consistently.

**Comment by pooyan-dadvand** [comment]
So giving a warning when the linear solver wants to resize is not the best idea. You have all these information inside and outside to issue a warning or error or just resize without saying nothing.

**Comment by loumalouomega** [comment]
> 
> 
> So giving a warning when the linear solver wants to resize is not the best idea. You have all these information inside and outside to issue a warning or error or just resize without saying nothing.

It was just an example, is the linear solve the one which decides what to do with that information. But the linear solver does not know exactly what happened. That's why I think the warning is important

File: kratos/linear_solvers/linear_solver.h [hunk 1]
```diff
@@ -257,48 +254,115 @@ class LinearSolver
...
+        KRATOS_WARNING_IF("LinearSolver", size !=  size_b) << "WARNING: Your RHS vector is not properly sized. Sizes: " << size << " vs " << size_b << std::endl ;
+        
+        return ((size ==  size_a) &&
+                (size ==  size_x) &&
+                (size ==  size_b));
     }
 
-    virtual bool IsConsistent(SparseMatrixType& rA, DenseMatrixType& rX, DenseMatrixType& rB)
+    /**
+     * @brief This method checks if the dimensions of the system of equations are consistent (dense matrix version)
```
Comment (philbucher) [inline kratos/linear_solvers/linear_solver.h]: here you say dense matrix version but the argument is a sparse matrix

**Comment by philbucher** [comment]
> > So giving a warning when the linear solver wants to resize is not the best idea. You have all these information inside and outside to issue a warning or error or just resize without saying nothing.
> 
> It was just an example, is the linear solve the one which decides what to do with that information. But the linear solver does not know exactly what happened. That's why I think the warning is important

if you say the linear solver needs this info then maybe an idea would be to return a `struct` containing the information needed?

**Comment by loumalouomega** [comment]
> 
> 
> > > So giving a warning when the linear solver wants to resize is not the best idea. You have all these information inside and outside to issue a warning or error or just resize without saying nothing.
> > 
> > 
> > It was just an example, is the linear solve the one which decides what to do with that information. But the linear solver does not know exactly what happened. That's why I think the warning is important
> 
> if you say the linear solver needs this info then maybe an idea would be to return a `struct` containing the information needed?

I was thinking more in a enum with the corresponding cases

File: kratos/linear_solvers/linear_solver.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/linear_solvers/linear_solver.h]: The rX and Rb are dense matrices instead of vectors

Comment (philbucher) [inline kratos/linear_solvers/linear_solver.h]:
```suggestion
     * @brief This method checks if the dimensions of the system of equations are consistent (dense vector version)
```

Comment (loumalouomega) [reply kratos/linear_solvers/linear_solver.h]: Technically is a dense matrix

File: kratos/linear_solvers/linear_solver.h [hunk 2]
```diff
@@ -257,48 +254,115 @@ class LinearSolver
...
+        VectorType& rB
+        )
     {
         return (!IsConsistent(rA, rX, rB));
     }
-
-    virtual bool IsNotConsistent(SparseMatrixType& rA, DenseMatrixType& rX, DenseMatrixType& rB)
+    
+    /**
+     * @brief This method checks if the dimensions of the system of equations are not consistent (dense matrix for RHS and unknowns version)
```
Comment (philbucher) [inline kratos/linear_solvers/linear_solver.h]: IMO this is wrong/misleading, the argument is a sparse matrix

```suggestion
     * @brief This method checks if the dimensions of the system of equations are not consistent
```

maybe omitting it?

**Comment by loumalouomega** [comment]
Ping

**Comment by pooyan-dadvand** [comment]
Still I don't got your usecase/reason for issuing the warning there....

**Comment by loumalouomega** [comment]
> 
> 
> Still I don't got your usecase/reason for issuing the warning there....

Everything should be done in the solver side?

**Comment by loumalouomega** [comment]
With the corresponding code duplication

**Comment by pooyan-dadvand** [comment]
I mean, if I have a method `foo` which wants to check the consistency:
```c++
foo(){
    KRATOS_ERROR_IF (mSolver.IsNotConsistent(A,x,b)) << "Not consistence system matrices << std::endl;
}
```
The same method would issue warning or error or simply call the resize depending on the task it does and the knowledge it has. Please note that you are sending the warning because, as you mentioned, the solver doesn't know exactly what happens. So it is not its responsibility to do it and we should not do it there.

**Comment by loumalouomega** [comment]
> 
> 
> I mean, if I have a method `foo` which wants to check the consistency:
> 
> ```c++
> foo(){
>     KRATOS_ERROR_IF (mSolver.IsNotConsistent(A,x,b)) << "Not consistence system matrices << std::endl;
> }
> ```
> 
> The same method would issue warning or error or simply call the resize depending on the task it does and the knowledge it has. Please note that you are sending the warning because, as you mentioned, the solver doesn't know exactly what happens. So it is not its responsibility to do it and we should not do it there.

Ok, now it is just a small clean up

**Review by pooyan-dadvand — APPROVED** [review]
:+1:

**Pull request merged**